Make sure PIP install fails fast and dump logs on failure

PIP install batches should fail fast otherwise it's unclear what is the nature of the failure:

```
[DEBUG] Executing command: '/Users/runner/work/pybuilder/pybuilder/.pybuilder/plugins/cpython-3.7.12.final.0/bin/python3' '-m' 'pip.__main__' 'install' '--upgrade' '--upgrade-strategy' 'only-if-needed' 'coverage~=6.0' 'coveralls~=2.1' 'flake8~=3.7' 'pdoc3>=0.8.3' 'pypandoc~=1.4' 'setuptools>=38.6.0' 'twine>=1.15.0' 'unittest-xml-reporting~=3.0.4' 'wheel>=0.34.0'
[DEBUG] Executing command: '/Users/runner/work/pybuilder/pybuilder/.pybuilder/plugins/cpython-3.7.12.final.0/bin/python3' '-m' 'pip.__main__' 'install' 'cram' 'sphinx'
------------------------------------------------------------
BUILD FAILED - Unable to install plugin packages into /Users/runner/work/pybuilder/pybuilder/.pybuilder/plugins/cpython-3.7.12.final.0. Please see '/Users/runner/work/pybuilder/pybuilder/.pybuilder/plugins/cpython-3.7.12.final.0/install.log' for full details:
	Collecting sphinxcontrib-jsmath
	  Downloading sphinxcontrib_jsmath-1.0.1-py2.py3-none-any.whl (5.1 kB)
	Collecting pytz>=2015.7
	  Downloading pytz-2021.3-py2.py3-none-any.whl (503 kB)
	Collecting MarkupSafe>=2.0
	  Downloading MarkupSafe-2.0.1-cp37-cp37m-macosx_10_9_x86_64.whl (13 kB)
	Collecting urllib3<1.27,>=1.21.1
	  Downloading urllib3-1.26.7-py2.py3-none-any.whl (138 kB)
	Collecting certifi>=2017.4.17
	  Downloading certifi-2021.10.8-py2.py3-none-any.whl (149 kB)
	Collecting idna<4,>=2.5
	  Downloading idna-3.2-py3-none-any.whl (59 kB)
	Collecting charset-normalizer~=2.0.0
	  Downloading charset_normalizer-2.0.7-py3-none-any.whl (38 kB)
	Collecting pyparsing>=2.0.2
	  Using cached pyparsing-2.4.7-py2.py3-none-any.whl (67 kB)
	Installing collected packages: urllib3, pytz, pyparsing, MarkupSafe, idna, charset-normalizer, certifi, sphinxcontrib-serializinghtml, sphinxcontrib-qthelp, sphinxcontrib-jsmath, sphinxcontrib-htmlhelp, sphinxcontrib-devhelp, sphinxcontrib-applehelp, snowballstemmer, requests, Pygments, packaging, Jinja2, imagesize, docutils, babel, alabaster, sphinx, cram
	Successfully installed Jinja2-3.0.2 MarkupSafe-2.0.1 Pygments-2.10.0 alabaster-0.7.12 babel-2.9.1 certifi-2021.10.8 charset-normalizer-2.0.7 cram-0.7 docutils-0.17.1 idna-3.2 imagesize-1.2.0 packaging-21.0 pyparsing-2.4.7 pytz-2021.3 requests-2.26.0 snowballstemmer-2.1.0 sphinx-4.2.0 sphinxcontrib-applehelp-1.0.2 sphinxcontrib-devhelp-1.0.2 sphinxcontrib-htmlhelp-2.0.0 sphinxcontrib-jsmath-1.0.1 sphinxcontrib-qthelp-1.0.3 sphinxcontrib-serializinghtml-1.1.5 urllib3-1.26.7
	WARNING: You are using pip version 21.2.4; however, version 21.3 is available.
	You should consider upgrading via the '/Users/runner/work/pybuilder/pybuilder/.pybuilder/plugins/cpython-3.7.12.final.0/bin/python3 -m pip install --upgrade pip' command. (pybuilder/install_utils.py:102)
------------------------------------------------------------
Build finished at 2021-10-12 02:23:20
Build took 19 seconds (19605 ms)
```

--- a/.committerrc
+++ b/.committerrc
@@ -1,6 +1,0 @@
-# committer configuration file
-# see https://github.com/aelgru/committer
-
-[COMMIT]
-# This command will be executed before perfoming a commit
-execute_before = ./build.py

--- a/.github/workflows/pybuilder.yml
+++ b/.github/workflows/pybuilder.yml
@@ -38,6 +38,7 @@ jobs:
       GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
     steps:
       - shell: bash
+        name: Configure
         if: |
           contains(env.DEPLOY_OSES, runner.os) &&
           contains(env.DEPLOY_PYTHONS, matrix.python-version) &&
@@ -46,6 +47,7 @@ jobs:
         run: |
           echo "PYB_EXTRA_ARGS='+upload'" >> $GITHUB_ENV
       - uses: pybuilder/build@master
+        name: Build
         if: |
           !(matrix.with-homebrew == 'true' && runner.os != 'macOS' ||
             matrix.with-homebrew == 'true' && runner.os == 'macOS' &&
@@ -62,6 +64,15 @@ jobs:
           install-pyb: 'false'
           pyb-command: $PYTHON build.py
           pyb-extra-args: +install ${{ env.PYB_EXTRA_ARGS }}
+      - shell: bash
+        name: Print Logs on Failure
+        if: failure()
+        run: |
+          for f in $(find . -name \*.log); do
+              echo ">>>>>> $f"
+              cat "$f"
+          done
+          exit 1
   build-stable-summary:
     if: success() || failure()
     runs-on: ubuntu-latest

--- a/src/main/python/pybuilder/install_utils.py
+++ b/src/main/python/pybuilder/install_utils.py
@@ -78,28 +78,31 @@ def install_dependencies(logger, project, dependencies, python_env,
             logger.warn("PIP will be operating in the offline mode")
 
         with open(np(log_file_name), log_file_mode) as log_file:
-            results = pip_install_batches(install_batch,
-                                          python_env,
-                                          index_url=project.get_property("install_dependencies_index_url"),
-                                          extra_index_url=project.get_property("install_dependencies_extra_index_url"),
-                                          trusted_host=project.get_property("install_dependencies_trusted_host"),
-                                          insecure_installs=project.get_property(
-                                              "install_dependencies_insecure_installation"),
-                                          verbose=project.get_property("pip_verbose"),
-                                          constraint_file=constraints_file,
-                                          logger=logger,
-                                          outfile_name=log_file,
-                                          error_file_name=log_file,
-                                          target_dir=target_dir,
-                                          ignore_installed=ignore_installed)
-        for result in results:
-            if result:
-                raise BuildFailedException("Unable to install %s packages into %s. "
-                                           "Please see '%s' for full details:\n%s",
-                                           package_type,
-                                           python_env.env_dir,
-                                           log_file_name,
-                                           tail_log(log_file_name))
+            for result in pip_install_batches(install_batch,
+                                              python_env,
+                                              index_url=project.get_property("install_dependencies_index_url"),
+                                              extra_index_url=project.get_property(
+                                                  "install_dependencies_extra_index_url"),
+                                              trusted_host=project.get_property("install_dependencies_trusted_host"),
+                                              insecure_installs=project.get_property(
+                                                  "install_dependencies_insecure_installation"),
+                                              verbose=project.get_property("pip_verbose"),
+                                              constraint_file=constraints_file,
+                                              logger=logger,
+                                              outfile_name=log_file,
+                                              error_file_name=log_file,
+                                              target_dir=target_dir,
+                                              ignore_installed=ignore_installed):
+                if result:
+                    try:
+                        log_file.close()
+                    finally:
+                        raise BuildFailedException("Unable to install %s packages into %s. "
+                                                   "Please see '%s' for full details:\n%s",
+                                                   package_type,
+                                                   python_env.env_dir,
+                                                   log_file_name,
+                                                   tail_log(log_file_name))
     return dependencies_to_install
 
 

--- a/src/main/python/pybuilder/pip_utils.py
+++ b/src/main/python/pybuilder/pip_utils.py
@@ -79,22 +79,19 @@ def pip_install_batches(packages, python_env, index_url=None, extra_index_url=No
             batches[opts] = batch_pkgs
         batch_pkgs.append(pkg_spec)
 
-    results = []
     for opts, pkgs in batches.items():
         cmd_line = list(pip_command_line)
         cmd_line.extend(opts)
         for pkg in pkgs:
             cmd_line.extend(pkg)
 
-        results.append(python_env.execute_command(cmd_line,
-                                                  outfile_name=outfile_name,
-                                                  error_file_name=error_file_name,
-                                                  env=env_environ,
-                                                  cwd=cwd,
-                                                  shell=False,
-                                                  no_path_search=True))
-
-    return results
+        yield python_env.execute_command(cmd_line,
+                                         outfile_name=outfile_name,
+                                         error_file_name=error_file_name,
+                                         env=env_environ,
+                                         cwd=cwd,
+                                         shell=False,
+                                         no_path_search=True)
 
 
 def pip_install(install_targets, python_env, index_url=None, extra_index_url=None, upgrade=False,

